File: benchmarks/lru_cache/java/Lru_cache.java
import java.util.HashMap;
import java.util.Map;

public class Lru_cache {
    static final int CAPACITY = 1000;
    static final long WORKING_SET = 800;

    static Map<Long, Integer> map = new HashMap<>();
    static long[] keyArr = new long[CAPACITY];
    static int[] prevArr = new int[CAPACITY];
    static int[] nextArr = new int[CAPACITY];
    static int head, tail, size;
    static long hitCount;

    static void listRemove(int slot) {
        int p = prevArr[slot];
        int nx = nextArr[slot];
        if (p != -1) nextArr[p] = nx; else head = nx;
        if (nx != -1) prevArr[nx] = p; else tail = p;
    }

    static void listPushFront(int slot) {
        prevArr[slot] = -1;
        nextArr[slot] = head;
        if (head != -1) prevArr[head] = slot;
        head = slot;
        if (tail == -1) tail = slot;
    }

    static void reset() {
        map = new HashMap<>();
        head = -1;
        tail = -1;
        size = 0;
        hitCount = 0;
    }

    static void accessKey(long key) {
        Integer existing = map.get(key);
        if (existing != null) {
            hitCount++;
            listRemove(existing);
            listPushFront(existing);
            return;
        }

        int slot;
        if (size < CAPACITY) {
            slot = size;
            size++;
        } else {
            slot = tail;
            listRemove(slot);
            map.remove(keyArr[slot]);
        }
        keyArr[slot] = key;
        map.put(key, slot);
        listPushFront(slot);
    }

    public static void main(String[] args) {
        long n = args.length > 0 ? Long.parseLong(args[0]) : 2_000_000;

        reset();
        for (long i = 0; i < CAPACITY + 50; i++) {
            accessKey(i);
        }
        if (map.containsKey(0L)) {
            System.err.println("self-check failed: key 0 should have been evicted");
            System.exit(1);
        }
        if (!map.containsKey((long) (CAPACITY + 49))) {
            System.err.println("self-check failed: most recently used key should still be cached");
            System.exit(1);
        }
        if (size != CAPACITY) {
            System.err.println("self-check failed: cache size should equal capacity after overfill");
            System.exit(1);
        }

        reset();
        for (long i = 0; i < n; i++) {
            accessKey(i % WORKING_SET);
        }

        long expected = n > WORKING_SET ? n - WORKING_SET : 0;
        if (hitCount != expected) {
            System.err.println("self-check failed: hit count mismatch: got " + hitCount + ", expected " + expected);
            System.exit(1);
        }

        System.out.println(hitCount);
    }
}

File: benchmarks/lru_cache/rust/lru_cache.rs
use std::collections::HashMap;
use std::env;

const CAPACITY: usize = 1000;
const WORKING_SET: i64 = 800;

struct Cache {
    map: HashMap<i64, usize>,
    key_arr: Vec<i64>,
    prev_arr: Vec<i32>,
    next_arr: Vec<i32>,
    head: i32,
    tail: i32,
    size: usize,
    hit_count: i64,
}

impl Cache {
    fn new() -> Self {
        Cache {
            map: HashMap::new(),
            key_arr: vec![0; CAPACITY],
            prev_arr: vec![-1; CAPACITY],
            next_arr: vec![-1; CAPACITY],
            head: -1,
            tail: -1,
            size: 0,
            hit_count: 0,
        }
    }

    fn list_remove(&mut self, slot: usize) {
        let p = self.prev_arr[slot];
        let nx = self.next_arr[slot];
        if p != -1 {
            self.next_arr[p as usize] = nx;
        } else {
            self.head = nx;
        }
        if nx != -1 {
            self.prev_arr[nx as usize] = p;
        } else {
            self.tail = p;
        }
    }

    fn list_push_front(&mut self, slot: usize) {
        self.prev_arr[slot] = -1;
        self.next_arr[slot] = self.head;
        if self.head != -1 {
            self.prev_arr[self.head as usize] = slot as i32;
        }
        self.head = slot as i32;
        if self.tail == -1 {
            self.tail = slot as i32;
        }
    }

    fn access_key(&mut self, key: i64) {
        if let Some(&slot) = self.map.get(&key) {
            self.hit_count += 1;
            self.list_remove(slot);
            self.list_push_front(slot);
            return;
        }

        let slot: usize;
        if self.size < CAPACITY {
            slot = self.size;
            self.size += 1;
        } else {
            slot = self.tail as usize;
            self.list_remove(slot);
            self.map.remove(&self.key_arr[slot]);
        }
        self.key_arr[slot] = key;
        self.map.insert(key, slot);
        self.list_push_front(slot);
    }
}

fn main() {
    let n: i64 = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(2_000_000);

    let mut cache = Cache::new();
    for i in 0..(CAPACITY as i64 + 50) {
        cache.access_key(i);
    }
    assert!(!cache.map.contains_key(&0), "self-check failed: key 0 should have been evicted");
    assert!(
        cache.map.contains_key(&(CAPACITY as i64 + 49)),
        "self-check failed: most recently used key should still be cached"
    );
    assert_eq!(cache.size, CAPACITY, "self-check failed: cache size should equal capacity after overfill");

    let mut cache = Cache::new();
    for i in 0..n {
        cache.access_key(i % WORKING_SET);
    }

    let expected = if n > WORKING_SET { n - WORKING_SET } else { 0 };
    assert_eq!(cache.hit_count, expected, "self-check failed: hit count mismatch");

    println!("{}", cache.hit_count);
}
